lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,453 @@
+				// Update the known tip to the newest one.
+				if tip_is_new {
+					// First check for any unconfirmed transactions and act on it immediately.
+					match self.get_unconfirmed_transactions(&confirmables) {
+						Ok(unconfirmed_txs) => {
+							// Double-check the tip hash. If it changed, a reorg happened since
+							// we started syncing and we need to restart last-minute.
+							{
+								let check_notification = self.client.block_headers_subscribe()?;
+								let check_tip_hash = check_notification.header.block_hash();
+
+								// Restart if either the tip changed or we got some divergent tip
+								// change notification since we started.
+								let mut restart_sync = check_tip_hash != tip_header.block_hash();
+								while let Some(queued_notif) = self.client.block_headers_pop()? {
+									if queued_notif.header.block_hash() != check_tip_hash {
+										restart_sync = true
+									}
+								}
+
+								if restart_sync {
+									tip_header = check_notification.header;
+									tip_height = check_notification.height as u32;
+									log_debug!(self.logger,
+										"Encountered inconsistency during transaction sync, restarting.");
+									sync_state.pending_sync = true;
+									continue;
+								}
+							}
+
+							num_unconfirmed += unconfirmed_txs.len();
+							sync_state.sync_unconfirmed_transactions(&confirmables,
+								unconfirmed_txs);
+						},
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							log_error!(self.logger,
+								"Failed during transaction sync, aborting. Synced so far: {} confirmed, {} unconfirmed.",
+								num_confirmed,
+								num_unconfirmed
+							);
+							sync_state.pending_sync = true;
+							return Err(TxSyncError::from(err));
+						}
+					}
+
+					// Update the best block.
+					for c in &confirmables {
+						c.best_block_updated(&tip_header, tip_height);
+					}
+				}
+
+				match self.get_confirmed_transactions(&sync_state) {
+					Ok(confirmed_txs) => {
+						// Double-check the tip hash. If it changed, a reorg happened since
+						// we started syncing and we need to restart last-minute.
+						{
+							let check_notification = self.client.block_headers_subscribe()?;
+							let check_tip_hash = check_notification.header.block_hash();
+
+							// Restart if either the tip changed or we got some divergent tip
+							// change notification since we started.
+							let mut restart_sync = check_tip_hash != tip_header.block_hash();
+							while let Some(queued_notif) = self.client.block_headers_pop()? {
+								if queued_notif.header.block_hash() != check_tip_hash {
+									restart_sync = true
+								}
+							}
+
+							if restart_sync {
+								tip_header = check_notification.header;
+								tip_height = check_notification.height as u32;
+								log_debug!(self.logger,
+									"Encountered inconsistency during transaction sync, restarting.");
+								sync_state.pending_sync = true;
+								continue;
+							}
```

Comment:
Not that critical, but we could DRY this up by having an abstraction for polling the chain tip.